src/zx_graph.jl

```diff
@@ -114,24 +120,58 @@ function add_edge!(zxg::ZXGraph, v1::Integer, v2::Integer, edge_type::Int = HADA
                 else
                     return false
                 end
-            else
-                return add_edge!(zxg.mg, v1, v2, edge_type)
+            elseif add_edge!(zxg.mg, v1, v2)
+                @inbounds zxg.et[(min(v1, v2), max(v1, v2))] = edge_type
+                return true
             end
         end
-    else
-        return false
     end
+    return false
 end
 
 spider_type(zxg::ZXGraph, v::Integer) = zxg.st[v]
 phase(zxg::ZXGraph, v::Integer) = zxg.ps[v]
-is_hadamard(e::MultipleEdge) = (mul(e) == HADAMARD)
-is_hadamard(zxg::ZXGraph, v1::Integer, v2::Integer) = (mul(zxg.mg, v1, v2) == HADAMARD)
+function set_phase!(zxg::ZXGraph{T, P}, v::T, p::P) where {T, P}
+    if has_vertex(zxg.mg, v)
+        while p < 0
+            p += 2
+        end
+        p = rem(p, one(P)+one(P))
+        @inbounds zxg.ps[v] = p
+        return true
+    end
+    return false
+end
+nqubits(zxg::ZXGraph) = zxg.layout.nbits
+
+qubit_loc(zxg::ZXGraph{T, P}, v::T) where {T, P} = qubit_loc(zxg.layout, v)
+function column_loc(zxg::ZXGraph{T, P}, v::T) where {T, P}
+    c_loc = column_loc(zxg.layout, v)
+    if c_loc !== nothing
+        if spider_type(zxg, v) == SpiderType.Out
+            nb = neighbors(zxg, v)[]
+            spider_type(zxg, nb) == SpiderType.In && return 3//1
+            c_loc = floor(column_loc(zxg, nb) + 2)
+        end
+        if spider_type(zxg, v) == SpiderType.In
+            nb = neighbors(zxg, v)[]
+            spider_type(zxg, nb) == SpiderType.Out && return 1//1
+            c_loc = ceil(column_loc(zxg, nb) - 2)
+        end
+    end
+    return c_loc
+end
+
+function is_hadamard(zxg::ZXGraph, v1::Integer, v2::Integer)
+    src = min(v1, v2)
+    dst = max(v1, v2)
+    @inbounds return zxg.et[(src, dst)] == EdgeType.HAD
+end
 spiders(zxg::ZXGraph) = vertices(zxg.mg)
 
 function rem_spiders!(zxg::ZXGraph{T, P}, vs::Vector{T}) where {T, P}
     if rem_vertices!(zxg.mg, vs)
-        for v in vs
+        @simd for v in vs
```

Comment:
same question, is `@simd` faster? need to benchmark it, it probably won't work here.